src/passes/Inlining.cpp

```diff
@@ -332,6 +357,432 @@ doInlining(Module* module, Function* into, const InliningAction& action) {
+
+      split.inlineable = copyFunction(func, "inlineable-A");
+      auto* outlined = copyFunction(func, "outlined-A");
+
+      // The inlineable function should only have the if, which will call the
+      // outlined function with a flipped condition.
+      auto* inlineableIf = getIf(split.inlineable->body);
+      inlineableIf->condition =
+        builder.makeUnary(EqZInt32, inlineableIf->condition);
+      inlineableIf->ifTrue = builder.makeCall(
+        outlined->name, getForwardedArgs(func, builder), Type::none);
+      split.inlineable->body = inlineableIf;
+
+      // The outlined function no longer needs the initial if.
+      //
+      // Note that the body must be a block, because if it were not then the
+      // function would be easily inlineable (just an if with a simple condition
+      // and a return), and we would not even attempt to do splitting.
+      auto& outlinedList = outlined->body->cast<Block>()->list;
+      outlinedList.erase(outlinedList.begin());
+
+      *inlineableOut = split.inlineable;
+      return true;
+    }
+
+    // Pattern B: Represents a function whose entire body looks like
+    //
+    //  if (A_1) {
+    //    ..heavy work..
+    //  }
+    //  ..
+    //  if (A_k) {
+    //    ..heavy work..
+    //  }
+    //  B; // an optional final value (which can be a return value)
+    //
+    // where there is a small number of such ifs with arguments A1..A_k, and
+    // A_1..A_k and B (if the final value B exists) are very simple.
+    //
+    // Also, each if body must either be unreachable, or it must have type none
+    // and have no returns. If it is unreachable, for example because it is a
+    // return, then we will just return a value from the outlined function:
+    //
+    //  if (A_i) {
+    //    return outlined(..);
+    //  }
+    //
+    // Or, if an if body has type none, then for now we assume that we do not
+    // need to return a value from there, which makes things simpler, and in
+    // that case we just do this, which continues onward in the function:
+    //
+    //  if (A_i) {
+    //    outlined(..);
+    //  }
+    //
+    // TODO: handle a possible returned value in this case as well.
+    //
+    // Note that the if body type must be unreachable or none, as this is an if
+    // without an else.
+
+    // Find the number of ifs.
+    // TODO: Investigate more values here. 4 appears useful on real-world code.
+    const Index MaxIfs = 4;
+    Index numIfs = 0;
+    while (getIf(body, numIfs) && numIfs <= MaxIfs) {
+      numIfs++;
+    }
+    if (numIfs == 0 || numIfs > MaxIfs) {
+      return false;
+    }
+
+    // Look for a final item after the ifs.
+    auto* finalItem = getItem(body, numIfs);
+
+    // The final item must be simple (or not exist, which is simple enough).
+    if (finalItem && !isSimple(finalItem)) {
+      return false;
+    }
+
+    // There must be no other items after the optional final one.
+    if (getItem(body, numIfs + (finalItem ? 1 : 0))) {
```

Comment:
If there is no final item, then we already know there is nothing after the `If`s:
```suggestion
    if (finalItem && getItem(body, numIfs + 1)) {
```